src/Data/Generic.purs

```diff
@@ -49,6 +51,24 @@ class Generic a where
     toSignature :: Proxy a -> GenericSignature
     fromSpine :: GenericSpine -> Maybe a
 
+-- | Checks that the spine follows the structure defined by the signature
+isValidSpine :: GenericSignature -> GenericSpine -> Boolean
+isValidSpine SigBoolean (SBoolean _) = true
+isValidSpine SigNumber (SNumber _) = true
+isValidSpine SigInt (SInt _) = true
+isValidSpine SigString (SString _) = true
+isValidSpine (SigArray sig) (SArray spines) = all (isValidSpine (sig unit) <<< (unit #)) spines
+isValidSpine (SigProd alts) (SProd tag values) =
+  case find ((tag ==) <<< _.sigConstructor) alts of
+    Nothing -> false
+    Just { sigValues: sigValues } ->
+      and $ zipWith (\sig spine -> isValidSpine (sig unit) (spine unit)) sigValues values
+isValidSpine (SigRecord fieldSigs) (SRecord fieldVals) =
+  and $ zipWith (\sig val -> isValidSpine (sig.recValue unit) (val.recValue unit))
+                (sortBy (\a b -> compare a.recLabel b.recLabel) fieldSigs)
```

Comment:
@paf31 That's exactly what I had originally, but `purescript-functions` doesn't seem to be in transitive dependencies and I thought that it's better not to pull it for such tiny thing.